--- /src/main/java/com/vaadin/data/fieldgroup/FieldGroup.java	2016-10-03 11:48:38.000000000 +0300
+++ /src/main/java/com/vaadin/data/fieldgroup/FieldGroup.java	2016-10-03 11:48:16.000000000 +0300
@@ -56,7 +56,7 @@
  */
 public class FieldGroup implements Serializable {
 
-    private Item itemDataSource;
+    protected Item itemDataSource;
     private boolean buffered = true;

     private boolean enabled = true;
